neurokit2/signal/signal_fixpeaks.py

```diff
@@ -563,67 +594,89 @@ def _plot_artifacts_lipponen2019(artifacts, info):
-    while continue_loop is True:
-        if interval_max is not None:
-            interval = signal_period(peaks, sampling_rate=sampling_rate, desired_length=None)
-            peaks, continue_loop = _interpolate_missing(
-                peaks, interval, interval_max, sampling_rate
-            )
-
-        if relative_interval_max is not None:
-            interval = signal_period(peaks, sampling_rate=sampling_rate, desired_length=None)
-            interval = standardize(interval, robust=robust)
-            peaks, continue_loop = _interpolate_missing(
-                peaks, interval, relative_interval_max, sampling_rate
-            )
-
+    if interval_max is not None:
+        interval = signal_period(
+            peaks, sampling_rate=sampling_rate, desired_length=None
+        )
+        peaks = _interpolate_missing(
+            peaks=peaks,
+            interval=interval,
+            interval_max=interval_max,
+            sampling_rate=sampling_rate,
+        )
+    if relative_interval_max is not None:
+        interval = signal_period(
+            peaks, sampling_rate=sampling_rate, desired_length=None
+        )
+        interval = standardize(interval, robust=robust)
+        peaks = _interpolate_missing(
+            peaks=peaks,
+            interval=interval,
+            interval_max=relative_interval_max,
+            sampling_rate=sampling_rate,
+        )
     return peaks
 
 
-def _interpolate_missing(peaks, interval, interval_max, sampling_rate):
+def _interpolate_missing(
+    peaks,
+    interval,
+    interval_max,
+    sampling_rate,
+):
     outliers = interval > interval_max
     outliers_loc = np.where(outliers)[0]
+
+    # interval returned by signal_period at index 0 is the mean of the intervals
+    # so it does not actually correspond to whether the first peak is an outlier
+    outliers_loc = outliers_loc[outliers_loc != 0]
+
     if np.sum(outliers) == 0:
-        return peaks, False
+        return peaks
+    peaks_to_correct = peaks.copy().astype(float)
 
-    # Delete large interval and replace by two unknown intervals
-    interval[outliers] = np.nan
-    interval = np.insert(interval, outliers_loc, np.nan)
-    #    new_peaks_location = np.where(np.isnan(interval))[0]
+    interval_without_outliers = interval[np.invert(outliers)]
 
-    # Interpolate values
-    interval = pd.Series(interval).interpolate().values
-    peaks_corrected = _period_to_location(interval, sampling_rate, first_location=peaks[0])
-    peaks = np.insert(
-        peaks, outliers_loc, peaks_corrected[outliers_loc + np.arange(len(outliers_loc))]
-    )
-    return peaks, True
+    # go through the outliers starting with the highest indices
+    # so that the indices of the other outliers are not moved when
+    # unknown intervas are inserted 
+    for loc in np.flip(outliers_loc):
+        # compute number of NaNs to insert based on the mean interval
+        n_nan = int(interval[loc] / np.nanmean(interval_without_outliers))
```

Comment:
One final nitpick 🙈 `np.nanmean(interval_without_outliers)` won't change throughout the iterations of the for loop. I.e., it's a loop-invariant that's unnecessarily re-computed on every iteration. You could assign it to a variable before you enter the loop, and then use the variable inside the loop (i.e., "hoist the loop-invariant").

Reply:
Yes good point, and thanks for teaching me the vocab :D